### ISSUE TITLE
Low performance on high-resolution canvases with OSD v5 and Firefox

### ISSUE THREAD

**Issue by zfox23** [openseadragon/openseadragon#2569]
I have deployed OSD on my website for viewing gigapixel photographs. The OSD canvas resizes upon page load and when the window is resized. When the browser window is maximized, the dimensions of the OSD canvas is nearly the same as the screen dimensions.

**When my Firefox browser window is large and thus the pixel dimensions of my OSD canvas are large, the framerate of the OSD canvas is low - perhaps 15fps.** This is not the case in OSD pre-v5.

When I shrink the Firefox window to be small (and thus decrease the dimensions of the canvas), the framerate of the canvas increases to my monitor's refresh rate, which is 144hz.

This problematic behavior does not occur in Chrome with OSD v5. When using Chrome, no matter the dimensions of the canvas, the framerate remains at 144hz.

Thanks for a wonderful piece of software!

## Additional Debugging Info
- I am testing on [this webpage](https://zachfox.photography/stories/gigapixel-manhattan-skyline-from-astoria-nyc/), which is running v4 (NOT v5) as of filing this issue.

### Hardware/Software
- i7-13900k
- RTX 4090
- 3840x1600px@144hz display
- Windows 11
- Firefox v129

**Comment by pearcetm** [comment]
Interesting behavior - thanks for testing! The big thing that has changed is that the viewer now uses `WebGL` as the default renderer, instead of `2D canvas`, but that isn't the only thing that is different between v4 and v5. Could you try using the option `drawer: 'canvas'` on Firefox with the big window? If performance of `canvas` is the same as in v4 and it is only `webgl` that takes the FPS hit, we'd at least know it is in fact something about Firefox's `webgl` implementation that is causing this.

**Comment by zfox23** [comment]
> Interesting behavior - thanks for testing! The big thing that has changed is that the viewer now uses `WebGL` as the default renderer, instead of `2D canvas`, but that isn't the only thing that is different between v4 and v5. Could you try using the option `drawer: 'canvas'` on Firefox with the big window? If performance of `canvas` is the same as in v4 and it is only `webgl` that takes the FPS hit, we'd at least know it is in fact something about Firefox's `webgl` implementation that is causing this.

Hi pearcetm! Yes, I'd be happy to test with the canvas drawer. However, I don't know how to do that. It looks like `drawer` is not an option on the `OpenSeadragon` constructor. Am I missing something?

**Comment by pearcetm** [comment]
You should be able to pass `drawer: 'canvas'` as an option: https://openseadragon.github.io/docs/OpenSeadragon.html#.Options

**Comment by iangilman** [comment]
@zfox23 Are you getting the same results on https://openseadragon.github.io/? Also, how are you testing the FPS?

**Comment by zfox23** [comment]
> @zfox23 Are you getting the same results on https://openseadragon.github.io/? Also, how are you testing the FPS?

@iangilman: Yes, I see the same results on https://openseadragon.github.io/ when I toggle fullscreen mode. Without toggling fullscreen mode, performance is fine.

I'm testing the framerate observationally, not using any tools or getting numeric data. The performance difference is immediately apparent. I could use Firefox's profiling tools to get more data, but I haven't done that yet.

> You should be able to pass `drawer: 'canvas'` as an option: https://openseadragon.github.io/docs/OpenSeadragon.html#.Options

Ah, I see! Thank you. The TypeScript definitions don't seem to contain that option, or I am missing something. I'll give that a shot ASAP.

**Comment by pearcetm** [comment]
> The TypeScript definitions don't seem to contain that option, or I am missing something.

The TypeScript definitions are from a different repo (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/openseadragon) - looks like that could use some updating.

**Comment by iangilman** [comment]
Good point on the TypeScript definitions! I've added a comment here: https://github.com/openseadragon/openseadragon/issues/1628#issuecomment-2293832119

For what it's worth, when I go to https://openseadragon.github.io/ in Firefox and go to fullscreen, the performance seems okay to me. It's definitely not 60 fps, but it's not annoyingly laggy. I would be curious to get an actual fps reading, but I'm not sure how best to do that. I'm on Firefox 129.0.1 on MacOS 14.5 on a reasonably powerful Mac.

**Comment by zfox23** [comment]
> You should be able to pass `drawer: 'canvas'` as an option: https://openseadragon.github.io/docs/OpenSeadragon.html#.Options

Okay! This worked, and the WebGL-based performance issues are gone. So - something's up with WebGL + OSD on Firefox 🤔

**Comment by zfox23** [comment]
Here's a Firefox performance profile I just captured from https://openseadragon.github.io/:

[Download Firefox 2024-08-16 21.28 profile.json.gz](https://github.com/user-attachments/files/16642980/Firefox.2024-08-16.21.28.profile.json.gz)

You can open this profile at https://profiler.firefox.com/.

In the first half of the profile, the OSD canvas is fullscreen, and I am panning around the image. In the second half of the profile, the OSD canvas is its original size on the page (i.e. relatively small), and I am panning around the image:

![image](https://github.com/user-attachments/assets/e09e377e-2748-4397-89ec-e3a341ae473e)

When the OSD canvas is fullscreen, it looks like Firefox is spending a lot of time in `CanvasRenderingContext2D.drawImage()`:

![image](https://github.com/user-attachments/assets/50076de9-6133-4f66-999e-59f4b909fa69)

I'm not sure what can change in OSD's code to make that more performant. Perhaps nothing? ☹️

As for raw framerate numbers: I'm trying to figure out how to get that data right now and will make a new comment if I figure it out.

**Comment by pearcetm** [comment]
If you maximize the window, and have the viewer take the full window (but
not actually fullscreen mode), is the performance closer to full-screen or
small viewer performance? Just trying to narrow down whether it is related
to canvas size in general or a special case for true full screen mode.

Googling “webgl performance Firefox fullscreen”, there are a number of
reports of issues in other web apps too (like YouTube), specifically in
this circumstance, with various suggestions about fixing user profiles etc.
to potentially help.

Since it is limited to Firefox but not Chrome and appears to affect other
apps too, I’m not sure if it’s fixable within OSD. If not, maybe we should
consider making webgl “opt-in” rather than default, @iangilman? Or at a
minimum add documentation about the issue? I think more info may be needed
and we shouldn’t necessarily rush to conclusions, just want to start
thinking about it…

(Since it appears to be browser specific it would be nice to have Firefox
fix it!)

**Comment by zfox23** [comment]
> If you maximize the window, and have the viewer take the full window (but not actually fullscreen mode), is the performance closer to full-screen or small viewer performance? Just trying to narrow down whether it is related to canvas size in general or a special case for true full screen mode. \<snip>

The performance in this case is poor (i.e. closer to fullscreen). Thus, I think the problem is related to canvas size.

_(The way my OSD viewer is set up on my website is exactly this case; the viewer takes up almost the entire window without needing to be in fullscreen mode, and performance is poor.)_

> (Since it appears to be browser specific it would be nice to have Firefox
fix it!)

I agree! I've had WebGL performance woes in Firefox with my own projects in the past. I couldn't do anything to fix those, either.

**Comment by iangilman** [comment]
Presumably the Firefox folks are aware they have WebGL performance issues. I don't know if it would help them to have OSD as an example scenario, but if someone wants to bring it to them, it can't hurt!

Meanwhile, sounds like maybe we need to work around it. There may indeed be things we could do to improve performance on Firefox. Perhaps some aspect of our pipeline is causing the majority of the slowdown. I guess it depends on whether anyone is up for digging into that.

Otherwise, it seems like adding an "auto" mode for the drawer that picks WebGL on Chrome and canvas on Firefox (and whatever is appropriate for other browsers) is probably the next step. And yeah, that should probably be the default. Of course we'll need to do more education for devs to make sure they know how to switch drawer modes depending on what plugins they need.

**Comment by iangilman** [comment]
I guess I'll add that another complication is perhaps the performance of the computer running Firefox. Like I said, mine seems to do just fine (on a 4k monitor). Of course detecting that is much more challenging than detecting whether you're on Firefox.

@zfox23 what OS are you on?

**Comment by zfox23** [comment]
> I guess I'll add that another complication is perhaps the performance of the computer running Firefox. Like I said, mine seems to do just fine (on a 4k monitor). Of course detecting that is much more challenging than detecting whether you're on Firefox.
> 
> @zfox23 what OS are you on?

True!

Hardware/software details:
- i7-13900k
- RTX 4090
- 3840x1600px@144hz display
- Windows 11
- Firefox v129

**Comment by iangilman** [comment]
Yeah, sounds pretty equivalent to my setup, just Windows instead of Mac. Interesting.

**Comment by pranavgoel29** [comment]
I’d like to add that I’m experiencing the same issue with Firefox. In my case, it's jitter in both non-fullscreen and fullscreen however the performance drops significantly when using full screen mode (both page and true full screen). Interestingly, everything runs smoothly on Edge, Chrome, and other Chromium-based browsers. I’ve tested this on both a Samsung Tab S10+ and an iPad Pro, and the issue persists across both devices.

**Comment by joedf** [comment]
I wanted to add that I am also experiencing a performance hit in Firefox with OpenSeadragon `v5.0.1`. In MS Edge / Chrome, it is all fine, always 60 FPS. In Firefox it can drop to 15-25 FPS when panning.

**System**
- CPU: Intel i9-11900H
- GPU 0: Intel(R) UHD Graphics (128 MB)
- GPU 1: NVIDIA GeForce RTX 3050 Ti Laptop GPU (4 GB)
- RAM: 32.0 GB
- OS: Windows 10
- Display: 3840 x 2400 px (250% scaling), 60 Hz
- Firefox v140.0.1

**Drawer Performance**
I am using [mrdoob's stats.js](http://github.com/mrdoob/stats.js) to measure this.
- Firefox (GPU 0)
  - `html`
    - home zoom: 60 FPS
    - 100% zoom: 59-60 FPS
  - `canvas`
    - home zoom: 25-38 FPS
    - 100% zoom: 19-21 FPS
  - `webgl`
    - home zoom: 22-24 FPS
    - 100% zoom: 19-22 FPS
- Firefox (GPU 1): all round sluggish, 12-18 FPS
- Microsoft Edge v137.0.3296.93
  - `html` and `canvas` (any zoom): 60 FPS
  - `webgl` (GPU 0):
    - home zoom: 44-45 FPS
    - 100% zoom: 34-38 FPS
  - `webgl` (GPU 1, any zoom): 60 FPS

**Notes**
Just wanted to make note that you can tell Windows what GPU to use with the Graphical Settings page.
![Image](https://github.com/user-attachments/assets/63f8b93f-780b-4fbb-b574-e9548ff28660)

I am using 1024x1024 px tiles.

I read somewhere in Firefox's bugzilla that using dedicated GPU / hardware-acceleration is still wonky in Firefox, and is as such, not the default GPU for this reason.

Interestingly, the 'html' drawer is insanely smooth and fast everywhere, but that is likely because it is CPU-bound and I am on an Intel Core i9.

**Possible bug with `requestDrawer`?**
It seems that if I do `viewer.requestDrawer('html')` only some of the tiles are loaded. If I zoom out, I can a blank viewer, if I zoom in some tiles are visible... Then if I programmatically reload the tile source, all the tiles show up fine. Switching to `canvas` or `webgl` all work fine.

**Comment by joedf** [comment]
> [...] Otherwise, it seems like adding an "auto" mode for the drawer that picks WebGL on Chrome and canvas on Firefox (and whatever is appropriate for other browsers) is probably the next step. [...]

Yeah, I think an `auto` mode like this makes the most sense. In my experience, in some weird obscure cases, Firefox is way faster than Chrome, but for the most part Chrome is seemingly faster.

**Comment by iangilman** [comment]
@joedf Thank you for testing this and all the data!

Sounds like we have a plan. We just need someone to implement it.

**Comment by FlyingDR** [comment]
It looks like the reason of the observable behavior is use of the [Resist fingerprinting](https://support.mozilla.org/en-US/kb/resist-fingerprinting) feature in Firefox.

When `privacy.resistFingerprinting` option is enabled in `about:config` OSD spends lots of time in `CanvasRenderingContext2D.drawImage` effectively causing high CPU usage:

<img width="731" height="348" alt="Image" src="https://github.com/user-attachments/assets/8d5ba35f-f745-462a-9b38-828b2aa0ee4f" />

<img width="735" height="276" alt="Image" src="https://github.com/user-attachments/assets/f95284ee-78c2-44e9-8bae-9d52418abae5" />

When this setting is turned off - there is no background activity and no CPU usage.

Hope this information will help to find a possible workaround.

**Comment by joedf** [comment]
Good catch! This makes a lot of sense to be the reason, since Firefox has more hardened fingerprinting prevention on by default which involves some randomization (as far as I know) in some js canvas and image element use-cases.

**Comment by iangilman** [comment]
Wow, yes, great catch! I wasn't aware of this feature.

Unfortunately I guess it's not something we have control over in the code (would defeat the purpose, after all), so it's just something we would have to educate users about.

**Comment by pearcetm** [comment]
> Wow, yes, great catch! I wasn't aware of this feature.
> 
> Unfortunately I guess it's not something we have control over in the code (would defeat the purpose, after all), so it's just something we would have to educate users about.

Agree - good catch, and I didn't know about this until reading about it here. It looks like not only is it something we can't control, there also are no good ways to detect whether it is on or not (also for security purposes).

@iangilman Do you have any particular kind of end user education in mind?

**Comment by zfox23** [comment]
Hi again! Issue filer here. I have another datapoint:

**`privacy.resistFingerprinting` is and has always been `false` in my browser**, which I believe is the default setting. There are other settings hierarchically underneath `privacy.resistFingerprinting`. which are enabled. I do not know if those settings apply even if the base `privacy.resistFingerprinting` is `false`.

Here's what my relevant settings look like. All of these settings are set to their defaults:
<img width="456" height="646" alt="Image" src="https://github.com/user-attachments/assets/dc2777bd-ce36-4c9b-894a-430d4e738300" />

I don't think this setting being enabled is the only reason why we'd see decreased performance in Firefox.

**Comment by joedf** [comment]
I think it depends... My tests were most likely with it on. Mainly, I have the "Enhanced Tracking Protection" on by default.
More about it here: https://support.mozilla.org/en-US/kb/enhanced-tracking-protection-firefox-desktop

**Comment by FlyingDR** [comment]
Here are some additional observations:

For experiments I've used image in DZI format with this statistics for tiles per level:

 - level 8: 1 tile
 - level 9: 2 tiles
 - level 10: 6 tiles
 - level 11: 15 tiles
 - level 12: 54 tiles
 - level 13: 204 tiles

Without `privacy.resistFingerprinting` turned on OSD loads tiles from levels 8..12 (**86 requests** in total), but with this setting turned on level 13 is **also being loaded**, effectively causing **339 tile requests** to be made. 

Since every tile load triggers call for `drawWorld()` function - it causes 4 times more redraws and it, of course consumes more time and causes CPU load.

This behavior caused by one of the features that are adjusted in this mode (reported in [this article](https://support.mozilla.org/en-US/kb/resist-fingerprinting) inside "What does Resist Fingerprinting alter?" section):

> The device pixel ratio is reported as 2.0

It means that `window.devicePixelRatio` returns `2`, so `getCurrentPixelDensityRatio()` returns possibly higher value and this value is used for most of calculations.

I did not dig deep in OSD logic, but it looks like some kind of additional heuristics for the world redraw requirement will certainly help.

Still, performance of the `CanvasRenderingContext2D.drawImage` seem to be the main reason, so it may worth to fill a bugreport for Firefox.

**Comment by iangilman** [comment]
Wild stuff! Good to know the resist fingerprinting is off by default. I'm guessing it's still in flux.

@FlyingDR If you're up for filing a bug report on Firefox, please do!
